File: Problem148_snakes-and-ladders.java
// Time Complexity: O(m*n)
// Space Complexity: O(m*n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach

class Solution {
    public int snakesAndLadders(int[][] board) {
        if(board == null || board.length == 0) return 0;
        int n = board.length;
        int[] moves = new int[n * n];
        int r = n - 1;
        int c = 0;
        // flatten the matrix
        int even = 0, idx = 0; // idx on moves
        while(r >= 0 && c >= 0) {
            if(board[r][c] == -1) {
                moves[idx] = -1;
            } else {
                moves[idx] = board[r][c] - 1; 
            }
            idx++;
            if(even % 2 == 0) {
                c++;
                if(c == n) {
                    r--;
                    c--;
                    even++;
                }
            } else {
                c--;
                if(c == -1) {
                    r--;
                    c++;
                    even++;
                }
            }
        }
        
        // traverse and add it to queue
        Queue<Integer> q = new LinkedList<>();
        q.add(0); 
        int count = 0;
        moves[0] = -2;
        while(!q.isEmpty()) {
            int size = q.size();
            for(int i = 0; i < size; i++) {
                int curr = q.poll();
                if(curr == n*n - 1) return count;
                for(int j = 1; j < 7; j++) {
                    int child = curr + j; // baby idx
                    if(child >= n*n) continue;
                    if(moves[child] != -2) {
                        if(moves[child] == -1) {
                            q.add(child);
                        } else {
                            q.add(moves[child]);
                        }
                        moves[child] = -2;
                    }
                }
            }
            count++;
        }
        return -1;
    }
}

File: Problem-148 Snakes and Ladders.py
# 909. Snakes and Ladders
# https://leetcode.com/problems/snakes-and-ladders/

# Logic: Flatten the board into a linear list. Start BFS 
# from first cell and do level order traversal. When you 
# cross the end or reach the end, return the count.

# Time Complexity: O(n*n)
# Space Complexity: O(n)

from collections import deque

class Solution:
    def snakesAndLadders(self, board) -> int:
        n = len(board)
        moves = [1,2,3,4,5,6]
        cells = list()
        flag = True
        
        for i in range(n-1, -1, -1):
            if not flag:
                cells += board[i][::-1]
            else:
                cells += board[i]
            flag = not flag
        
        for idx, i in enumerate(cells):
            if cells[idx] == -1:
                cells[idx] = idx + 1

        visited = set()
        q = deque()
        count = 0
        
        visited.add(cells[0])
        q.append(cells[0])
        
        while q:
            size = len(q)
            count += 1

            for i in range(size):
                cur = q.popleft()
                
                for j in moves:
                    newCell = cur + j - 1
                    
                    if cells[newCell] >= (n**2):
                        return count
                    
                    if newCell < n**2 and cells[newCell] not in visited:
                        visited.add(cells[newCell])
                        q.append(cells[newCell])
                
        return -1

obj = Solution()
print(obj.snakesAndLadders([[-1,-1,-1,-1,-1,-1],[-1,-1,-1,-1,-1,-1],[-1,-1,-1,-1,-1,-1],[-1,35,-1,-1,13,-1],[-1,-1,-1,-1,-1,-1],[-1,15,-1,-1,-1,-1]]))
print(obj.snakesAndLadders([[-1,-1],[-1,3]]))